src/provider/store/Actions.ts

```diff
@@ -13,47 +14,91 @@ export const enum Action {
     TOGGLE_VISIBILITY = '@@ui/TOGGLE_CENTER_WINDOW',
 }
 
-export interface CreateNotification extends ReduxAction<Action> {
-    type: Action.CREATE,
-    notification: StoredNotification;
+export class BaseAction<T extends Action> implements ReduxAction<Action> {
+    public readonly type: T;
+
+    constructor(type: T) {
+        this.type = type;
+    }
+}
+
+export abstract class CustomAction<T extends Action> extends BaseAction<T> {
+    public abstract async dispatch(store: StoreAPI): Promise<void>;
+}
+
+export class CreateNotification extends CustomAction<Action.CREATE> {
+    public readonly notification: StoredNotification;
+
+    constructor(notification: StoredNotification) {
+        super(Action.CREATE);
+        this.notification = notification;
+    }
+
+    public async dispatch(store: StoreAPI): Promise<void> {
+        const notification = this.notification;
+        const existingNotifications = store.state.notifications.filter(x => x.id === notification.id);
+
+        if (existingNotifications.length) {
+            await store.dispatch(new RemoveNotifications(existingNotifications));
```

Comment:
Do we need both this and the code in the ActionHandlerMap for replacing a notification if we try to create a notification with the same ID?

Reply:
This is so we can send a 'closed' event to the client when a new notification with the same id is created. It also allows the toast manager to remove the old notification.

Reply:
Oh sure, but the `if (index >= 0) {` on line 103 will always be false, right? Because we've already modified storage in the `Action.REMOVE` action handler?

Reply:
As discussed, given the importance of these ID uniqueness constraints, will leave additional checks as-is - but will flag them as so, to make clear that these constraints exist.